packages/tools/client-debugger/client-debugger-chrome-extension/src/react-components/AudienceView.tsx

```diff
@@ -25,12 +34,58 @@ export interface AudienceViewProps extends HasContainerId {
 export function AudienceView(props: AudienceViewProps): React.ReactElement {
 	const { containerId } = props;
 
-	// TODO: Post message requesting Audience summary
-	// TODO: Listen for Audience data updates
+	const messageRelay = React.useContext(MessageRelayContext);
```

Comment:
Not related to this PR, but to kick off the discussion. @Josmithr the idea is that the client-debugger-view will also end up  using the context, right? So both versions end up handling messages the same, and at that point we can use the same "main" views (not just the underscore ones) in both places? I was also thinking if there might be advantages to instead exposing an abstraction of the debugger to feed into the views, and then an implementation for the devtools panel debugger can use the message passing mechanism and the implementation for the page-rendered debugger could go straight to the containers.

Reply:
My thinking hadn't gotten quite this far, but I think these are worth considering. I do think the more programmatic, event-based API is a much better abstraction to expose from the core debugger library perspective, I see the message passing as more of a necessary evil to support the chrome extension world. So I would be in favor of keeping both options functional from the debugger's perspective.

As for how the view library uses it, we definitely have a lot of duplication that could be cleaned up. There is something appealing about just using the simple event model where we can, but if our primary use case is the chrome extension, then it may not be worth all of the duplication required.

I also like the idea of being a bit more opt-in to what functionality lights up (e.g. events vs message-passing), the main problem from my perspective is that the debugger library doesn't necessarily know how it's being consumed, so we don't know what to light up at initialization time. But we could do something like expose an API for enabling different state communication mechanisms on demand (or something). The client-debugger-view package could (for example) expressly enable event-based reports on launch, and the chrome extension could pass a message requesting activation of message-passing-based reports (caveat here being that we would need to unconditionally listen for this activation message, regardless of whether or not the message-passing mode has been enabled yet).

This all definitely merits further discussion and investigation.